core/src/main/java/org/geysermc/geyser/bedrock/camera/GeyserCameraFade.java

```diff
@@ -0,0 +1,95 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.bedrock.camera;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.geysermc.geyser.api.bedrock.camera.CameraFade;
+
+import java.awt.Color;
+
+public record GeyserCameraFade(
+        Color color,
+        float fadeInSeconds,
+        float holdSeconds,
+        float fadeOutSeconds
+) implements CameraFade {
+
+    public static class CustomFadeBuilder implements CameraFade.Builder {
+        private Color color;
+        private float fadeInSeconds;
+        private float holdSeconds;
+        private float fadeOutSeconds;
+
+        @Override
+        public CameraFade.Builder color(@NonNull Color color) {
+            this.color = color;
+            return this;
+        }
+
+        @Override
+        public CameraFade.Builder fadeInSeconds(float fadeInSeconds) {
+            if (fadeInSeconds > 10f) {
+                throw new IllegalArgumentException("Fade in seconds must be at most 10 seconds");
+            }
+            this.fadeInSeconds = fadeInSeconds;
+            return this;
+        }
+
+        @Override
+        public CameraFade.Builder holdSeconds(float holdSeconds) {
+            if (holdSeconds > 10f) {
+                throw new IllegalArgumentException("Hold seconds must be at most 10 seconds");
+            }
+            this.holdSeconds = holdSeconds;
+            return this;
+        }
+
+        @Override
+        public CameraFade.Builder fadeOutSeconds(float fadeOutSeconds) {
+            if (fadeOutSeconds > 10f) {
+                throw new IllegalArgumentException("Fade out seconds must be at most 10 seconds");
+            }
+            this.fadeOutSeconds = fadeOutSeconds;
+            return this;
+        }
+
+        @Override
+        public CameraFade build() {
+            if (color == null) {
+                throw new IllegalArgumentException("Color must be set");
+            }
+
+            if (fadeInSeconds + holdSeconds + fadeOutSeconds < 0.5f) {
+                throw new IllegalArgumentException("Fade times must be at least 0.5 seconds");
+            }
+
+            if (fadeInSeconds + holdSeconds + fadeOutSeconds > 10f) {
+                throw new IllegalArgumentException("Fade times must be at most 10 seconds");
+            }
```

Comment:
I would reword it to something a bit more clear along the lines of "Total fade time (in, hold, out) must be..."